planner/core/plan_cost_ver2.go

```diff
@@ -0,0 +1,293 @@
+	rowSize := math.Max(p.getScanRowSize(), 2.0)
+	logRowSize := math.Log2(rowSize)
+
+	p.planCost = rows * logRowSize * scanFactor
+	p.planCostInit = true
+	return p.planCost, nil
+}
+
+// plan-cost = rows * log2(row-size) * scan-factor
+// log2(row-size) is from experiments.
+func (p *PhysicalTableScan) getPlanCostVer2(taskType property.TaskType, option *PlanCostOption) (float64, error) {
+	rows := getCardinality(p, option.CostFlag)
+	scanFactor := getTaskScanFactor(p, taskType)
+	rowSize := math.Max(p.getScanRowSize(), 2.0)
+	logRowSize := math.Log2(rowSize)
+
+	p.planCost = rows * logRowSize * scanFactor
+
+	// give TiFlash a start-up cost to let the optimizer prefers to use TiKV to process small table scans.
+	if p.StoreType == kv.TiFlash {
+		p.planCost += 2000 * logRowSize * scanFactor
+	}
+
+	p.planCostInit = true
+	return p.planCost, nil
+}
+
+// plan-cost = (child-cost + net-cost + seek-cost) / concurrency
+// net-cost = rows * row-size * net-factor
+// seek-cost = num-tasks * seek-factor
+func (p *PhysicalIndexReader) getPlanCostVer2(taskType property.TaskType, option *PlanCostOption) (float64, error) {
+	rows := getCardinality(p.indexPlan, option.CostFlag)
+	rowSize := getTblStats(p.indexPlan).GetAvgRowSize(p.ctx, p.indexPlan.Schema().Columns, true, false)
+	netFactor := getTaskNetFactor(p, taskType)
+	concurrency := float64(p.ctx.GetSessionVars().DistSQLScanConcurrency())
+
+	netCost := rows * rowSize * netFactor
+	seekCost := estimateNetSeekCost(p.indexPlan)
+
+	childCost, err := p.indexPlan.GetPlanCost(property.CopSingleReadTaskType, option)
+	if err != nil {
+		return 0, err
+	}
+
+	p.planCost = (childCost + netCost + seekCost) / concurrency
+	p.planCostInit = true
+	return p.planCost, nil
+}
+
+// plan-cost = (child-cost + net-cost + seek-cost) / concurrency
+// net-cost = rows * row-size * net-factor
+// seek-cost = num-tasks * seek-factor
+func (p *PhysicalTableReader) getPlanCostVer2(taskType property.TaskType, option *PlanCostOption) (float64, error) {
+	rows := getCardinality(p.tablePlan, option.CostFlag)
+	rowSize := getTblStats(p.tablePlan).GetAvgRowSize(p.ctx, p.tablePlan.Schema().Columns, true, false)
+	netFactor := getTaskNetFactor(p, taskType)
+	concurrency := float64(p.ctx.GetSessionVars().DistSQLScanConcurrency())
+
+	netCost := rows * rowSize * netFactor
+	seekCost := estimateNetSeekCost(p.tablePlan)
+
+	childCost, err := p.tablePlan.GetPlanCost(property.CopSingleReadTaskType, option)
+	if err != nil {
+		return 0, err
+	}
+
+	p.planCost = (childCost + netCost + seekCost) / concurrency
+	p.planCostInit = true
+
+	// consider tidb_enforce_mpp
+	_, isMPP := p.tablePlan.(*PhysicalExchangeSender)
+	if isMPP && p.ctx.GetSessionVars().IsMPPEnforced() &&
+		!hasCostFlag(option.CostFlag, CostFlagRecalculate) { // show the real cost in explain-statements
+		p.planCost /= 1000000000
+	}
+	return p.planCost, nil
+}
+
+// plan-cost = index-side-cost + (table-side-cost + double-read-cost) / double-read-concurrency
```

Comment:
We'd better follow the format that the first comment begins with `funcName ...`

Reply:
Updated, PTAL.